mesa/experimental/cell_space.py

```diff
@@ -0,0 +1,247 @@
+import itertools
+import random
+from collections.abc import Iterable
+from functools import cache, cached_property
+from typing import Callable, Optional
+
+from .. import Agent
+
+Coordinate = tuple[int, int]
+
+
+class Cell:
+    __slots__ = ["coordinate", "_connections", "agents", "capacity", "properties"]
+
+    def __init__(self, i: int, j: int, capacity: int | None = 1) -> None:
+        self.coordinate = (i, j)
+        self._connections: list[Cell] = []
```

Comment:
See the discussion elsewhere with @rht, but what about turning this into a dict with the direction of the connection as keys, and the cells as values? So `self._connections["left"]` would return the cell to my left? You can do all kinds of funky stuff with that. it makes row and column retrieval trivial. But you can also express more complex moves (e.g., a knight move from chess) very concisely if you have the direction information on the connections somewhere.

Reply:
I would encode the left-right, down-up in a "Python turtle" agent's move methods, as an interpretation to the connection indices. This interpretation is specific to a 2D rectangular grid. While a cell collection signifying a geographical map wouldn't always have exactly 4 neighbors at all times (west/east neighbor, south/north neighbor), but sometimes would have 1 neighbor (Vatican cell and Italy cell).

Reply:
I think I understand your point. If we decide to go down this route, naming the connections (i.e., the keys in my proposal) should be handled in the `_connect_single_cell`. For a 2D grid and HexGrid, this is relatively easy and unambiguous. With networks or geographical maps, it becomes a bit more difficult.

Reply:
I am sympathetic towards the idea of named connections, but I think it's not required for a first version. Connections is a private variable so we change the data type to a dict later. The reason is that I think finding good names is not so trivial. For example for a knights move is it up, up, left or right, right, up, or any of the other possibilities? It would be unambiguous from an agents point of view (forward, forward, left or right), but i don't see an immediate translation to the directions yet. 

But in principle the idea is super cool, you would even be able to name some long-range connections between distant cells.
Also networks-on-grid would be easy

Reply:
Ok, let's leave this for later. 

I guess that there most be some proper terminology from the cellular automata world that we should be able to reuse.